### ISSUE TITLE
Recursion Error in 3.0.5-3.0.9: streamline Forward expression

### ISSUE THREAD

**Issue by austinmatherne-wk** [pyparsing/pyparsing#447]
I'm working on updating [an application](https://github.com/Arelle/Arelle/) from an old version of pyparsing. In the process I've run into recursion errors starting with 3.0.5 and continuing with the latest release (3.0.9). I've narrowed down the cause to the change to [this line](https://github.com/pyparsing/pyparsing/commit/9987004c94ccf7d9b6b3adbcf06d05d2ff197737#diff-daba53cec7bed1be7b180ee5e8378c772408d07afe6f2cad6d62e966993b9e45R38) in commit https://github.com/pyparsing/pyparsing/commit/9987004c94ccf7d9b6b3adbcf06d05d2ff197737. The gist of our usage is an instance of [Forward](https://github.com/pyparsing/pyparsing/blob/2c648a1acdd0fc6691b682dff1a23ce9b4005090/pyparsing/core.py#L5245) passed to [delimited_list](https://github.com/pyparsing/pyparsing/blob/2c648a1acdd0fc6691b682dff1a23ce9b4005090/pyparsing/helpers.py#L20) which with the change to that commit calls `streamline()` on the instance of Forward. 

I'm not that familiar with pyparsing or what streamline is supposed to do, but the streamline implementation for the Forward expression looks suspect. 

`self.expr` [can be reassigned](https://github.com/pyparsing/pyparsing/blob/2c648a1acdd0fc6691b682dff1a23ce9b4005090/pyparsing/core.py#L5288) in `__lshift__`, but it's checking `self.streamlined` to determine if `self.expr.streamline()` should be called.

It seems that either:
1. `self.expr.streamline()` should be called as long as `self.expr` is not `None` and the streamline implementation in `self.expr` should determine what to do if it's already been called.
2. `self.streamlined` should be reset in `__lshift__` when `self.expr` is reassigned.

https://github.com/pyparsing/pyparsing/blob/2c648a1acdd0fc6691b682dff1a23ce9b4005090/pyparsing/core.py#L5438-L5443

https://github.com/pyparsing/pyparsing/blob/2c648a1acdd0fc6691b682dff1a23ce9b4005090/pyparsing/core.py#L5279-L5298

**Comment by ptmcg** [comment]
Thanks for the detailed research! I'll try to repro with a Forward and delimitedList.

Thanks for putting in the effort with pyparsing!

**Comment by ptmcg** [comment]
I've not been able to reproduce this issue. Here is the code I wrote to try and repro, how does this compare with your code?

``` python
# https://github.com/pyparsing/pyparsing/issues/447
import string

import pyparsing as pp
pp.ParserElement.inline_literals_using(pp.Suppress)
print(pp.__version_info__)

part = pp.nested_expr()
part = pp.Forward().set_name("part")
part <<= pp.Group('(' + (pp.Char(string.ascii_letters) | part)[...] + ')')
partlist = pp.delimited_list(part)
# part <<= pp.Group('(' + (pp.Char(string.ascii_letters) | part)[...] + ')')

# try explicit streamline call
partlist.streamline()

# try printing expression
print(partlist)

# try creating diagram
# partlist.create_diagram("issue_447_diagram.html")

# try calling parse_string (called internally by run_tests)
partlist.run_tests("""\
    (a b c)
    (d(e f))
    (a b c),(d(e f))
    """)
```

**Comment by austinmatherne-wk** [comment]
I can confirm that works (that's [what we're doing](https://github.com/Arelle/Arelle/blob/ca4b7693868ccfaf33820cac7e546f78543ff4ef/arelle/XPathParser.py#L583-L588) as a workaround).

I haven't had any luck reproducing this in a simple example. The code that exercises this is quite complex (XPath parser written 10+ years ago running against XBRL formula specification conformance suite), but I'll keep trying to get you a more straightforward example.

**Comment by ptmcg** [comment]
Released in 3.1.0a1.
